wallet/src/Pos/Wallet/Web/Pending/Worker.hs

```diff
@@ -0,0 +1,167 @@
+{-# LANGUAGE Rank2Types   #-}
+{-# LANGUAGE TypeFamilies #-}
+
+-- | Pending transactions resubmition logic.
+
+module Pos.Wallet.Web.Pending.Worker
+    ( startPendingTxsResubmitter
+    ) where
+
+import           Universum
+
+import           Control.Monad.Catch          (Handler (..), catches)
+import           Data.Time.Units              (Second, convertUnit)
+import           Formatting                   (build, sformat, shown, stext, (%))
+import           Mockable                     (delay, fork)
+import           Serokell.Util.Text           (listJson)
+import           System.Wlog                  (logInfo, modifyLoggerName)
+
+import           Pos.Client.Txp.Addresses     (MonadAddresses)
+import           Pos.Communication            (submitAndSaveTx)
+import           Pos.Communication.Protocol   (SendActions (..))
+import           Pos.Constants                (pendingTxResubmitionPeriod)
+import           Pos.Core                     (FlatSlotId, SlotId (..), getBlockCount)
+import           Pos.Core.Context             (HasCoreConstants)
+import           Pos.Core.Slotting            (flattenSlotId)
+import           Pos.Crypto                   (WithHash (..))
+import           Pos.Slotting                 (getLastKnownSlotDuration, onNewSlot)
+import           Pos.Txp                      (ToilVerFailure (..), TxAux (..),
+                                               topsortTxs)
+import           Pos.Wallet.Web.Mode          (MonadWalletWebMode)
+import           Pos.Wallet.Web.Pending.Types (PendingTx (..), PtxCondition (..))
+import           Pos.Wallet.Web.Pending.Util  (isReclaimableFailure)
+import           Pos.Wallet.Web.State         (casPtxCondition, getPendingTxs)
+import           Pos.Wallet.Web.Util          (getWalletAssuredDepth)
+
+type MonadPendings m =
+    ( MonadWalletWebMode m
+    , MonadAddresses m
+    , HasCoreConstants
+    )
+
+-- | What should happen with existing pending transaction if attempt
+-- to resubmit it failed.
+processPtxFailure :: MonadPendings m => PendingTx -> ToilVerFailure -> m ()
+processPtxFailure PendingTx{..} e =
+    unless (isReclaimableFailure e) $ do
+        let newCond = PtxWontApply (sformat build e)
+        void $ casPtxCondition _ptxWallet _ptxTxId PtxApplying newCond
+        logInfo $ sformat ("Pending transaction "%build%" was canceled")
+                  _ptxTxId
+
+processPtxInUpperBlocks :: MonadPendings m => SlotId -> PendingTx -> m ()
+processPtxInUpperBlocks curSlot PendingTx{..} = do
+    mdepth <- getWalletAssuredDepth _ptxWallet
+    if | PtxInUpperBlocks slotId <- _ptxCond,
+         Just depth <- mdepth,
+         longAgo depth slotId -> do
+             void $ casPtxCondition _ptxWallet _ptxTxId _ptxCond PtxPersisted
+             logInfo $ sformat ("Transaction "%build%" got persistent") _ptxTxId
```

Comment:
are we going to store all persistent txs for the whole period of life of a node?